contrib/grpc-contrib/src/main/java/com/salesforce/grpc/contrib/interceptor/DebugClientInterceptor.java

```diff
@@ -0,0 +1,119 @@
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+
+/**
+ * {@code DebugClientInterceptor} intercepts the both outbound requests and inbound response to
+ * log the method name,status,headers and message as per given logging level.
+ *
+ */
+public class DebugClientInterceptor implements ClientInterceptor {
+    private final Logger logger = LoggerFactory.getLogger(DebugClientInterceptor.class);
+    private static final String REQUEST = "Request";
+    private static final String RESPONSE = "Response";
+
+    public enum Level {
+        NONE, STATUS, HEADERS, MESSAGE
+    }
+
+    private Level level = Level.NONE;
+
+    public DebugClientInterceptor(Level level) {
+        this.level = level;
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
+            CallOptions callOptions, Channel next) {
+        return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+            @Override
+            public void sendMessage(ReqT message) {
+                logMessage(REQUEST, message);
+                super.sendMessage(message);
+            }
+
+            @Override
+            public void start(Listener<RespT> responseListener, Metadata headers) {
+                logMethod(method);
+                logHeaders(REQUEST, headers);
+                super.start(
+                        new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(responseListener) {
+                            /**
+                             * for logging level of {@code HEADERS} or {@code MESSAGE}, the status will be
+                             * logged after headers and message
+                             */
+                            @Override
+                            public void onClose(Status status, Metadata trailers) {
+                                logStatus(status, method);
+                                super.onClose(status, trailers);
+                            }
+
+                            @Override
+                            public void onHeaders(Metadata headers) {
+                                logHeaders(RESPONSE, headers);
+                                super.onHeaders(headers);
+                            }
+
+                            @Override
+                            public void onMessage(RespT message) {
+                                logMessage(RESPONSE, message);
+                                super.onMessage(message);
+                            }
+                        }, headers);
+            }
+        };
+    }
+
+    /**
+     * override this method to change the way outbound request method name is logged.
+     */
```

Comment:
The API for overriding how printing works could use improvement. In the way this is written, users have to override four methods to get the desired effect. Also, the overrides don't have access to `level` because it is private, so they can't reimplement the conditional logic present in this class themselves.

A better option is to make each `logX()` method private, and have all four delegate to a shared `protected void log(String)` method to print out the result. That way, the user can override one method to modify printing.

Reply:
@rmichela if I get it right, I must construct the string message in each `logX()` and call log(String) in each of them, where log(String) would simply print the given string.Please correct I  am wrong.

Reply:
That's correct.